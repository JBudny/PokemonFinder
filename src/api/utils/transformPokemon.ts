import { Pokemon } from '@api/types/Pokemon'

import { PokemonItem } from '../types/pokemonItem.types'

interface TransformedPokemon {
	data: PokemonItem
}

export const transformPokemon = (pokemon: Pokemon): TransformedPokemon => {
	const { abilities, id, name, sprites, stats } = pokemon

	return {
		data: {
			abilities: abilities
				? abilities.map((ability) => ability.ability.name).join(', ')
				: '',
			id: id.toString(),
			name,
			stats,
			sprite: sprites.other['official-artwork'].front_default
		}
	}
}
