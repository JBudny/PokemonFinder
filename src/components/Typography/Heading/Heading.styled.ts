import { Text } from 'react-native'

import styled from 'styled-components/native'

import { TextSize } from '@components/Typography/Typography.types'

export const StyledHeading = styled(Text)<TextSize>`
	font-size: ${({ size, theme }) =>
		(size === 'large' && `${theme.designSystem.headingLargeSize}px`) ||
		(size === 'medium' && `${theme.designSystem.headingMediumSize}px`) ||
		(size === 'small' && `${theme.designSystem.headingSmallSize}px`)};
	font-weight: ${({ theme }) => theme.designSystem.headingWeight};
	color: ${({ theme }) => theme.palette.secondary};
`
