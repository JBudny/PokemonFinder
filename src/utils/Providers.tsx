import React, { FC } from 'react'
import { Provider } from 'react-redux'

import { ThemeProvider } from 'styled-components/native'

import { AppStore, setupStore } from '@src/store'
import { schema } from '@theme/schema'

const originalStore = setupStore()

interface ProvidersProps {
	store: AppStore
}

export const ProvideTheme: FC = ({ children }) => (
	<ThemeProvider theme={schema.light}>{children}</ThemeProvider>
)

export const Providers: FC<ProvidersProps> = ({
	children,
	store = originalStore
}) => (
	<Provider store={store}>
		<ProvideTheme>{children}</ProvideTheme>
	</Provider>
)
