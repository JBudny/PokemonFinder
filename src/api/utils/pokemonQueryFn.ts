import { BaseQueryApi } from '@reduxjs/toolkit/dist/query/baseQueryTypes'

import { transformPokemon } from './transformPokemon'

/**
 * Custom query function for the Pokemon API with error handling behavior
 * and data transformation logic.
 */
export const pokemonQueryFn = async (
	nameArg: string,
	_queryApi: BaseQueryApi,
	_extraOptions: Record<string, unknown>,
	fetchWithBQ: any
) => {
	const { data } = await fetchWithBQ(`/pokemon/${nameArg.toLowerCase()}`)

	try {
		return transformPokemon(data)
	} catch (e) {
		return {
			error: {
				status: 404,
				statusText: 'Pokemon not found!',
				data: undefined
			}
		}
	}
}
