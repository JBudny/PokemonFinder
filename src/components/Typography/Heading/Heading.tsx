import React, { FC } from 'react'

import { TypographyProps } from '@components/Typography/Typography.types'

import { StyledHeading } from './Heading.styled'

/**
 * Heading component based on Text core component.
 * @param size
 * @param text
 * EXTENDS: TextProps
 */
export const Heading: FC<TypographyProps> = ({ size, text, ...props }) => (
	<StyledHeading size={size} {...props}>
		{text}
	</StyledHeading>
)
