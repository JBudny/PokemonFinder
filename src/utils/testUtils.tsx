import React from 'react'

import type { PreloadedState } from '@reduxjs/toolkit'
import type { RenderOptions } from '@testing-library/react-native'
import { render } from '@testing-library/react-native'

import type { AppStore, RootState } from '@src/store'
import { setupStore } from '@src/store'

import { Providers } from './Providers'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store. For
// future dependencies, such as wanting to test with react-router, you can extend
// this interface to accept a path and route and use those in a <MemoryRouter />
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
	preloadedState?: PreloadedState<RootState>
	store?: AppStore
}

const renderWithProviders = (
	ui: React.ReactElement,
	{
		preloadedState = {},
		store = setupStore(preloadedState),
		...renderOptions
	}: ExtendedRenderOptions = {}
) => {
	const Wrapper: React.FC = ({ children }) => (
		<Providers store={store}>{children}</Providers>
	)

	return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

export { renderWithProviders }
