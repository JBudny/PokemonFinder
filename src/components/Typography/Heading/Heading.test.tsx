import React from 'react'

import { renderWithProviders } from '@src/utils/testUtils'

import { Heading } from './Heading'

test('Heading component should display provided text', () => {
	const { getByText } = renderWithProviders(
		<Heading size="small" text="test text" />
	)

	expect(getByText('test text')).toBeDefined()
})

test('Heading component with small text prop should return 20dp size text', () => {
	const { getByText } = renderWithProviders(
		<Heading size="small" text="test text" />
	)
	expect(getByText('test text')).toHaveStyle({ fontSize: 20 })
})

test('Heading component with medium text prop should return 25dp size text', () => {
	const { getByText } = renderWithProviders(
		<Heading size="medium" text="test text" />
	)
	expect(getByText('test text')).toHaveStyle({ fontSize: 25 })
})

test('Heading component with large text prop should return 35dp size text', () => {
	const { getByText } = renderWithProviders(
		<Heading size="large" text="test text" />
	)
	expect(getByText('test text')).toHaveStyle({ fontSize: 35 })
})

test('Heading component should match snapshot', () => {
	const text = renderWithProviders(<Heading size="small" text="test text" />)

	expect(text).toMatchSnapshot()
})
