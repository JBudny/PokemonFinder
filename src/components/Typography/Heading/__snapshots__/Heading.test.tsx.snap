// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component should match snapshot 1`] = `
<Text
  size="small"
  style={
    Array [
      Object {
        "color": "rgb(255, 255, 255)",
        "fontSize": 20,
        "fontWeight": "bold",
      },
    ]
  }
>
  test text
</Text>
`;
